xugrid/constants.py

```diff
@@ -22,6 +22,10 @@
 PolygonArray = np.ndarray
 SparseMatrix = Union[coo_matrix, csr_matrix]
 
+# Internally we always use a fill value of -1. This ensures we can always index
+# with the fill value as well, since any array will have at least size 1.
+FILL_VALUE = -1
```

Comment:
Indexing with the fill value is unavoidable because we are dealing with rectangular numpy arrays, rather than ragged/awkward arrays: https://github.com/scikit-hep/awkward

Let's say you have a mixed quad/triangular grid:

```python
vertices = np.array(
    [
        [0.0, 0.0],  # 0
        [1.0, 0.0],  # 1
        [2.0, 0.0],  # 2
        [0.0, 1.0],  # 3
        [1.0, 1.0],  # 4
        [2.0, 1.0],  # 5
        [1.0, 2.0],  # 6
    ]
)
faces = np.array(
    [
        [0, 1, 4, 3],
        [1, 2, 5, 4],
        [3, 4, 6, -1],
        [4, 5, 6, -1],
    ]
)
```

If you want the coordinates for every face, you simply do: `coords = vertices[faces]`.
This inserts node -1 (i.e. 6) for the triangles.
But we simply "remove" them with: `coords[faces == FILL_VALUE] = np.nan`.

Note that this allocates just one temporary array (`faces == FILL_VALUE`).

For reference, this is how you'd avoid indexing with a fill value:

```python
_faces = faces[faces != FILL_VALUE]  # 1D result
_coords = vertices[_faces]  # 1D result
coords = np.full((*faces.shape, 2), np.nan)
coords[faces == FILL_VALUE, :] = _coords
```

This allocates 4 temporaries: `_faces, _coords, faces != FILL_VALUE, faces == FILL_VALUE`

Using 0 temporarily:

```python
_faces = faces.copy()
_faces[face == FILL_VALUE] = 0
coords = vertices[_faces]
coords[face == FILL_VALUE] = np.nan
```

Doesn't help you one a bit: if the array is empty, it will IndexError all the same.

This is one example, but there are several others where having -1 is very convenient given that we're working with vectorized numpy operations.

Note that any valid Ugrid1d topology has atleast two vertices, and any valid Ugrid2d topology at least three vertices. Hence, -1 will always work.

Worrying about empty topologies is valid, but the check belongs at Ugrid1d/Ugrid2d initialization.

Reply:
If you're feeling clever / cute, you could add `(np.nan, np.nan)` as the last vertex. But the UGRID conventions don't specifiy that, they just specifiy a fill value. A fill value of -1 is fully conforming to UGRID conventions.

Reply:
Thanks for the explanation, I was under the misimpression that the FILL_VALUE was also used on places beyond the connectivity matrices. Probably should have read the UGRID conventions on this.